### ISSUE TITLE
[JENKINS-7046] Block Build when downstream project is building

### ISSUE THREAD

**Issue by jenkins-infra-bot** [timja-org/jenkins-gh-issues-poc-11-15#8361]
<p>When a downstream build is in progress, Hudson should prevent the upstream builds from occurring until the downstream build is complete.</p>

<p>For example:</p>

<p>CompA->CompB->CompC->CompD</p>

<p>If component C is building Hudson should block builds on A and B, until C has finished.  </p>

<p>It should then rebuild A, B and C before allowing D to proceed.</p>

<p>As the current system stands, if a user presses "build now" on component A when component C is building, A builds immediately.  This can be worked around in part by using the locks & latches plugin, but this has the side affect of disabling parallel Matrix builds which massively increases the build time (see <a href="https://issues.jenkins.io/browse/JENKINS-2725" title="locks and latches shouldnt' be global in multi-config" class="issue-link" data-issue-key="JENKINS-2725"><del>JENKINS-2725</del></a>).  Further, it should NOT be necessary to use ANY plugins to provide this fundamental behaviour. </p>

---
<details><summary><i>Originally reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=aggamemnon">aggamemnon</a>, imported from: <a href="https://issues.jenkins.io/browse/JENKINS-7046" target="_blank">Block Build when downstream project is building</a></i></summary>
<i><ul>
<li><b>status</b>: Resolved
<li><b>priority</b>: Major
<li><b>component(s)</b>: core
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2011-01-12T09:36:47+00:00
<li><b>votes</b>: 21
<li><b>watchers</b>: 20
<li><b>imported</b>: 2025-11-16
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>When a downstream build is in progress, Hudson should prevent the upstream builds from occurring until the downstream build is complete.</p>

<p>For example:</p>

<p>CompA-&gt;CompB-&gt;CompC-&gt;CompD</p>

<p>If component C is building Hudson should block builds on A and B, until C has finished.  </p>

<p>It should then rebuild A, B and C before allowing D to proceed.</p>

<p>As the current system stands, if a user presses "build now" on component A when component C is building, A builds immediately.  This can be worked around in part by using the locks &amp; latches plugin, but this has the side affect of disabling parallel Matrix builds which massively increases the build time (see <a href="https://issues.jenkins.io/browse/JENKINS-2725" title="locks and latches shouldnt' be global in multi-config" class="issue-link" data-issue-key="JENKINS-2725"><del>JENKINS-2725</del></a>).  Further, it should NOT be necessary to use ANY plugins to provide this fundamental behaviour. </p></pre>
</details>
</details>
<ul><li><i>environment</i>: <code>Dependant matrix builds across Linux and Solaris</code></li></ul>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>willemv's <a href="https://issues.jenkins.io/browse/JENKINS-7046?focusedId=142027&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-142027">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>After re-reading you original description, I can also confirm the bug : the 'Block build when upstream project is building' does not work transitively.</p>

<p>Ie. in the example of the reporter, A is not blocked when C is building, because A does not directly depend on C, only transitively.</p>

<p>We also observe this in our setup. Our hudson build is 1.376</p>
</pre>
</details>

</details>
<p>After re-reading you original description, I can also confirm the bug : the 'Block build when upstream project is building' does not work transitively.</p>

<p>Ie. in the example of the reporter, A is not blocked when C is building, because A does not directly depend on C, only transitively.</p>

<p>We also observe this in our setup. Our hudson build is 1.376</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/mindless.png" title="mindless's avatar" /> mindless's <a href="https://issues.jenkins.io/browse/JENKINS-7046?focusedId=142110&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-142110">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>Seems like 2 different things are discussed here.. willemv's last comment refers to "block build when upstream project is building" not working transitively.. I have also observed this behavior.  willemv, maybe file a new issue for this?  <a href="https://issues.jenkins.io/browse/JENKINS-5125" title="Projects building at the same time despite &quot;Block build when upstream is building&quot; option" class="issue-link" data-issue-key="JENKINS-5125">JENKINS-5125</a> and <a href="https://issues.jenkins.io/browse/JENKINS-5150" title="Race condition for feature &quot;Block build when upstream project is building&quot;" class="issue-link" data-issue-key="JENKINS-5150">JENKINS-5150</a> don't seem to specifically mention this..</p>

<p>However, it appears to me like the original request describes blocking in the opposite direction.. "block build when downstream project is building".  Hudson currently has no such feature.  Can you describe why you want this?  It doesn't make sense to me..  If we start the upstream job A, we know that will eventually trigger the downstream job C.. but since C has already started, there's nothing we can do at this point to avoid another build.  So why block A at this point?  Might as well run it in parallel with C to get all the builds completed more quickly.</p>
</pre>
</details>

</details>
<p>Seems like 2 different things are discussed here.. willemv's last comment refers to "block build when upstream project is building" not working transitively.. I have also observed this behavior.  willemv, maybe file a new issue for this?  <a href="https://issues.jenkins.io/browse/JENKINS-5125" title="Projects building at the same time despite "Block build when upstream is building" option" class="issue-link" data-issue-key="JENKINS-5125">JENKINS-5125</a> and <a href="https://issues.jenkins.io/browse/JENKINS-5150" title="Race condition for feature "Block build when upstream project is building"" class="issue-link" data-issue-key="JENKINS-5150">JENKINS-5150</a> don't seem to specifically mention this..</p>

<p>However, it appears to me like the original request describes blocking in the opposite direction.. "block build when downstream project is building".  Hudson currently has no such feature.  Can you describe why you want this?  It doesn't make sense to me..  If we start the upstream job A, we know that will eventually trigger the downstream job C.. but since C has already started, there's nothing we can do at this point to avoid another build.  So why block A at this point?  Might as well run it in parallel with C to get all the builds completed more quickly.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>heymjo's <a href="https://issues.jenkins.io/browse/JENKINS-7046?focusedId=142278&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-142278">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>(i took the liberty to update the subject to make it less confusing)</p>
</pre>
</details>

</details>
<p>(i took the liberty to update the subject to make it less confusing)</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>dmichonneau's <a href="https://issues.jenkins.io/browse/JENKINS-7046?focusedId=143197&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-143197">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>Here is a simple use case:</p>

<p>A: compilation of a jar
B: test the jar</p>

<p>B is running and locks the jar file
A is triggered: it fails to write the jar file after compilation.</p>
</pre>
</details>

</details>
<p>Here is a simple use case:</p>

<p>A: compilation of a jar<br/>
B: test the jar</p>

<p>B is running and locks the jar file<br/>
A is triggered: it fails to write the jar file after compilation.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/mindless.png" title="mindless's avatar" /> mindless's <a href="https://issues.jenkins.io/browse/JENKINS-7046?focusedId=143210&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-143210">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>For above use case I recommend using copyartifact plugin to copy the jar file to the workspace of job B, and use that when running tests.  Job A can now run while B is already running for a previous build of the jar, and A/B are also more flexible in that they can run on different slaves.</p>
</pre>
</details>

</details>
<p>For above use case I recommend using copyartifact plugin to copy the jar file to the workspace of job B, and use that when running tests.  Job A can now run while B is already running for a previous build of the jar, and A/B are also more flexible in that they can run on different slaves.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>dmichonneau's <a href="https://issues.jenkins.io/browse/JENKINS-7046?focusedId=143218&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-143218">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>That only moves the problem.</p>

<p>The jar copy is either done by A or B.</p>

<p>If done by A while B is running: the copy fails as B locks the file
If done by B while A is running: the copy fails as A locks the file. There is a shorter interval during which the file is locked, but it can still happen.</p>

<p>Besides that, it's an intrusive approach on tests that are not designed to run on something else that the compilation output directory.</p>
</pre>
</details>

</details>
<p>That only moves the problem.</p>

<p>The jar copy is either done by A or B.</p>

<p>If done by A while B is running: the copy fails as B locks the file<br/>
If done by B while A is running: the copy fails as A locks the file. There is a shorter interval during which the file is locked, but it can still happen.</p>

<p>Besides that, it's an intrusive approach on tests that are not designed to run on something else that the compilation output directory.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/mindless.png" title="mindless's avatar" /> mindless's <a href="https://issues.jenkins.io/browse/JENKINS-7046?focusedId=143620&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-143620">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<blockquote><p>If done by A while B is running: the copy fails as B locks the file</p></blockquote>

<p>I don't see how a running B can lock an artifact from a previous build.  Hudson makes a copy of the file when it archives the artifact.. it wouldn't be copying a file that is in-use/locked by a newly running build of B, right?</p>

<blockquote><p>If done by B while A is running: the copy fails as A locks the file. There is a shorter interval during which the file is locked, but it can still happen.</p></blockquote>

<p>B would never invoke a copy when A is configured with a copyartifact build step.</p>

<p>Here is more detail of what I was describing above:</p>
<ol>
	<li>A runs; Hudson archives an artifact</li>
	<li>B runs; it has a copyartifact build step to copy the artifact from the latest completed build of A.. if A is currently running that is ok; it is copying from the archived artifacts, not the workspace that is in use by that running build where files may be locked.</li>
	<li>A runs; if B is currently running that is ok since it is using a copy of a previous artifact.. this build can run to completion and archive a new artifact, then trigger a new build of B.</li>
</ol>


<p>If you have a "compilation output directory" instead of a simple artifact, I can see how this is less than ideal.</p>
</pre>
</details>

</details>
<blockquote><p>If done by A while B is running: the copy fails as B locks the file</p></blockquote>

<p>I don't see how a running B can lock an artifact from a previous build.  Hudson makes a copy of the file when it archives the artifact.. it wouldn't be copying a file that is in-use/locked by a newly running build of B, right?</p>

<blockquote><p>If done by B while A is running: the copy fails as A locks the file. There is a shorter interval during which the file is locked, but it can still happen.</p></blockquote>

<p>B would never invoke a copy when A is configured with a copyartifact build step.</p>

<p>Here is more detail of what I was describing above:</p>
<ol>
	<li>A runs; Hudson archives an artifact</li>
	<li>B runs; it has a copyartifact build step to copy the artifact from the latest completed build of A.. if A is currently running that is ok; it is copying from the archived artifacts, not the workspace that is in use by that running build where files may be locked.</li>
	<li>A runs; if B is currently running that is ok since it is using a copy of a previous artifact.. this build can run to completion and archive a new artifact, then trigger a new build of B.</li>
</ol>


<p>If you have a "compilation output directory" instead of a simple artifact, I can see how this is less than ideal.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/mindless.png" title="mindless's avatar" /> mindless's <a href="https://issues.jenkins.io/browse/JENKINS-7046?focusedId=143921&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-143921">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>Maybe it's worth prototyping a setup where you zip up the output directory and archive that zip as an artifact, in case this issue is not resolved for some time.. though maybe it's very large and that's why you haven't tried that already?</p>
</pre>
</details>

</details>
<p>Maybe it's worth prototyping a setup where you zip up the output directory and archive that zip as an artifact, in case this issue is not resolved for some time.. though maybe it's very large and that's why you haven't tried that already?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>aggamemnon's <a href="https://issues.jenkins.io/browse/JENKINS-7046?focusedId=144013&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-144013">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>"Maybe it's worth prototyping a setup where you zip up the output directory and archive that zip as an artifact, in case this issue is not resolved for some time.. though maybe it's very large and that's why you haven't tried that already?"</p>

<p>Not feasible due to the large size of the build directories, the large number of them and the extra unnecessary complexity this would add to the build environment.</p>
</pre>
</details>

</details>
<p>"Maybe it's worth prototyping a setup where you zip up the output directory and archive that zip as an artifact, in case this issue is not resolved for some time.. though maybe it's very large and that's why you haven't tried that already?"</p>

<p>Not feasible due to the large size of the build directories, the large number of them and the extra unnecessary complexity this would add to the build environment.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/mindless.png" title="mindless's avatar" /> mindless's <a href="https://issues.jenkins.io/browse/JENKINS-7046?focusedId=144034&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-144034">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>Have you tried a setup with locks&amp;latches plugin?</p>
</pre>
</details>

</details>
<p>Have you tried a setup with locks&latches plugin?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>mleinart's <a href="https://issues.jenkins.io/browse/JENKINS-7046?focusedId=144376&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-144376">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>Here's a working implementation: 
<a href="https://github.com/mleinart/hudson/tree/JENKINS-7046" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/mleinart/hudson/tree/JENKINS-7046</a></p>

<p>It's basically copy/paste/substitute from the upstream option.  A refactoring, especially of the display of the option  (pulldown of Upstream, Downstream, Both?) may be desired.</p>
</pre>
</details>

</details>
<p>Here's a working implementation: <br/>
<a href="https://github.com/mleinart/hudson/tree/JENKINS-7046" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/mleinart/hudson/tree/JENKINS-7046</a></p>

<p>It's basically copy/paste/substitute from the upstream option.  A refactoring, especially of the display of the option  (pulldown of Upstream, Downstream, Both?) may be desired.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>aggamemnon's <a href="https://issues.jenkins.io/browse/JENKINS-7046?focusedId=144617&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-144617">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>Fantastic, any idea on when this will make it into a release?</p>
</pre>
</details>

</details>
<p>Fantastic, any idea on when this will make it into a release?</p>
